src/inmanta/agent/agent.py

```diff
@@ -874,37 +885,54 @@ async def prepare_resources(
         """
         started = datetime.datetime.now().astimezone()
 
-        executor, invalid_resource_types = await self.setup_executor(model_version, resource_types)
+        executor, invalid_resources = await self.setup_executor(model_version, resource_types)
 
+        invalid_resources_types: dict[ResourceType, Exception] = {e.resource_type: e.exception for e in invalid_resources}
         loaded_resources: list[ResourceDetails] = []
-        failed_resources: list[ResourceVersionIdStr] = []
+        failed_resource_ids: list[ResourceVersionIdStr] = []
         undeployable: dict[ResourceVersionIdStr, const.ResourceState] = {}
 
+        logs = []
         for res in resource_batch:
-            if res["resource_type"] not in invalid_resource_types:
+            res_id = res["id"]
+            res_type = res["resource_type"]
+
+            if res_type not in invalid_resources_types:
                 loaded_resources.append(ResourceDetails(res))
 
                 state = const.ResourceState[res["status"]]
                 if state in const.UNDEPLOYABLE_STATES:
-                    undeployable[res["id"]] = state
+                    undeployable[res_id] = state
             else:
-                failed_resources.append(res["id"])
-                undeployable[res["id"]] = const.ResourceState.unavailable
+                logs.append(
```

Comment:
Isn't this going to repeat the same message over and over again?

I.e. We duplicate it for every instance of the type. 

I would just log the exception once per type (as we only produce one resource_action_update with all log messages at once)